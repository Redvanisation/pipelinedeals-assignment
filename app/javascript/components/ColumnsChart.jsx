import React from 'react';
import { Bar, defaults } from 'react-chartjs-2';
import PropTypes from 'prop-types';

defaults.global.legend.display = false;
defaults.global.defaultFontFamily = "'PT Sans', sans-serif"


const ColumnsChart = ({ isLoading, chartData }) => {


  return (
    <div>
      <h2 className="main-title">Column chart displaying total $ deal value in each stage</h2>
      {
        !isLoading
          ? (
            <Bar
              className="bar"
              data={chartData}
              options={{}}
           />
          )
          : <h2 className="loading">Content is loading!</h2>
      }
      
    </div>
  );
};

ColumnsChart.defaultProps = {
  isLoading: true
}

ColumnsChart.propTypes = {
  isLoading: PropTypes.bool,
  chartData: PropTypes.instanceOf(Object).isRequired
}

export default ColumnsChart;
