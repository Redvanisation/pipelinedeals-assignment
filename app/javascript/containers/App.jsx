import React, { useState, useEffect } from 'react';
import axios from 'axios';
import ColumnsChart from '../components/ColumnsChart';

const App = () => {

  const [isLoading, setIsLoading] = useState(true);
  const [chartData, setChartData] = useState([]);


  const setColors = (arr) => {
    let i = 0;
    const colors = [];
    while (i < arr.length) {
      colors.push(`rgba(${Math.floor(Math.random()*255)}, 
                        ${Math.floor(Math.random()*255)}, 
                        ${Math.floor(Math.random()*255)}, 
                    .7)`);
      i++;
    }
    return colors;
  }

  const getChart = () => {
    axios.get('/api/v1/deals')
    .then(res => {
      const values = [];
      const stages = Object.keys(res.data);
      let groupValue = 0;

      for (let obj in res.data) {
        for (let entry in res.data[obj]) {
          groupValue += res.data[obj][entry].value;
        }        
        values.push(groupValue);
      }

        setChartData({
          labels: stages, // stages
          datasets: [
            {
              data: values, // values
              backgroundColor: setColors(values),
            },
          ],
          options: {
            responsive: true,
            maintainAspectRatio: false,
          }
        });

        setIsLoading(false);
      })
      .catch(() => console.log('Error getting the data!'));
  };

  useEffect(() => {
    getChart();
  }, [])

  return (
    <ColumnsChart isLoading={isLoading} chartData={chartData} />
  );
}

export default App;
