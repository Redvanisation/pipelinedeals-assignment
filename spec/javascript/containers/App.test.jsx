import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

import App from '../../../app/javascript/containers/App';
import ColumnsChart from '../../../app/javascript/components/ColumnsChart';

describe('App container', () => {
  let wrapper;
  beforeEach(() => { wrapper = shallow(<App />) });

  it('renders correctly', () => {
    expect(wrapper).not.toBeNull();
  });

  it('Matches snapshot',  () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Has the columns chart component', () => {
    expect(wrapper.find(ColumnsChart)).toHaveLength(1);
  });

})