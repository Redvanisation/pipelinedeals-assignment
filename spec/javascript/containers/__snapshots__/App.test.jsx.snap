// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App container Matches snapshot 1`] = `ShallowWrapper {}`;
